### PR TITLE
Improving local variable debugging experience on optimized code

### PR DESCRIPTION
**Pull request by BrianBohe**
I am reporting local variables now and the integration has been checked from VS IDE. I am opening this PR to start iterating the solution, but there are changes needed:

NO MORE DESIRED:
1. Move instantiation of VariableLiveRange to when LclVarDsc is been created. After moving this outside Compiler class, I don't see the point of not initializing the complete array on genInitialize before iterating blocks and generating code inside genCodeForBBList.

DONE:
1. Only store the home's history of the variables that are in the IL (now I am tracking all the variables used in the linear scan register allocation).
2. See what changes are done with the stack level when the stackframe is not using the ebp and if this is not affecting one of my offsets. **Response:** it seems I don't have to care about that, information si always true on LclVarDsc at the moment of beign born/dead.
3. Review the changes applied on "psiScopeList" after genCodeForBBList. I have read there are some changes on the prolog and I am not updating offset info of args nor variables. Most of the code is inside a flag which by default is disabled (harcoded) on code, so I will not add code there.
4. Move VariableLiveRanges structures/classes outside compiler/LclVarDsc classes

TODO:
1. Measure Time/Memory performance changes of the solution
2. Add documentation
4. Add at least a test to the project so the functionality is not affected with next changes.

OUT OF SCOPE:
1. See how scopes are reported for promoted structs. Are they considered in the amount of local variables to report? The struct's attributes would not have a IL matching, so, Do I need to care about?

The items are commutative and the list is not complete.
Please fill free to send reviews.

### CODE REVIEW

**Comment by CarolEidt** [comment]
> 9. See how scopes are reported for promoted structs. Are they considered in the amount of local variables to report? The struct's attributes would not have a IL matching, so, Do I need to care about?

I doubt that promoted structs are reported to any degree of accuracy. My guess is that we simply fail to report them at all. I believe that adding support for these would be a non-trivial amount of work to change the scope info.

> Measure Time/Memory performance changes of the solution

In my opinion, this should be a second-order concern.

File: src/jit/codegenlinear.cpp [hunk 1]
```diff
@@ -808,12 +815,22 @@ void CodeGen::genSpillVar(GenTree* tree)
...
     varDsc->lvRegNum = REG_STK;
     if (varTypeIsMultiReg(tree))
     {
         varDsc->lvOtherReg = REG_STK;
     }
+
+    if (needsSpill)
+    {
+        // We need this after "lvRegNum" has change because now we are sure that varDsc->lvIsInReg() is false.
+        // "SiVarLoc" constructor uses the "LclVarDsc" of the variable.
```
Comment (CarolEidt) [inline src/jit/codegenlinear.cpp]: This comment seems a little strange here because there's no clear connection between the following and `SiVarLoc` or its constructor. I would just omit this line.

File: src/jit/codegenlinear.cpp [hunk 2]
```diff
@@ -1691,7 +1712,7 @@ void CodeGen::genConsumeBlockOp(GenTreeBlk* blkNode, regNumber dstReg, regNumber
 
 //-------------------------------------------------------------------------
 // genProduceReg: do liveness update for register produced by the current
-// node in codegen.
+// node in codegen once it was emitted.
```
Comment (CarolEidt) [inline src/jit/codegenlinear.cpp]: I would reword this to "after code has been emitted for it."

File: src/jit/compiler.cpp [hunk 3]
```diff
@@ -11091,3 +11091,182 @@ bool Compiler::killGCRefs(GenTree* tree)
...
+            amountOfLiveRanges += (unsigned int)varDsc->getAmountLiveRanges();
+        }
+    }
+
+    return amountOfLiveRanges;
+}
+
+//------------------------------------------------------------------------
+// startOrCloseVariableLiveRange: Starts or ends a "VariableLiveRange" for "varDsc" if it is
+// borning or dying respectively. Ranges are close-open "[)" so nothing is done in case
```
Comment (CarolEidt) [inline src/jit/compiler.cpp]: Wording: use "being born" in place of "borning" or "being borning", and "isBorn" in place of "isBorning". See `CodeGenInterface::genUpdateRegLife` in codegencommon.cpp

File: src/jit/compiler.h [hunk 4]
```diff
@@ -297,6 +297,89 @@ enum RefCountState
...
+                      emitLocation               startEmitLocation,
+                      emitLocation               endEmitLocation)
+    {
+        this->varLocation       = varLocation;
+        this->startEmitLocation = startEmitLocation;
+        this->endEmitLocation   = endEmitLocation;
+    }
+
+#if DEBUG
+    // Dump just the emitLocation as they are, we dont have generated the whole method yet
```
Comment (CarolEidt) [inline src/jit/compiler.h]: "dont have" should be "haven't"

File: src/jit/compiler.h [hunk 5]
```diff
@@ -319,6 +402,165 @@ class LclVarDsc
     {
     }
 
+    LiveRangeList* variableLiveRanges;
+
+#if DEBUG
+    LiveRangeBarrier* variableLifeBarrier;
```
Comment (CarolEidt) [inline src/jit/compiler.h]: What is this?

File: src/jit/compiler.h [hunk 6]
```diff
@@ -7092,6 +7338,20 @@ class Compiler
         return codeGen->getEmitter();
     }
 
+    // Starts or ends a "VariableLiveRange" for "varDsc" if it is borning or dying respectively.
+    // Ranges are close-open [) so nothing is done in case of being borning and dying at the same line
+    // due to be an empty range.
+    void startOrCloseVariableLiveRange(const LclVarDsc* varDsc, bool isBorning, bool isDying);
```
Comment (CarolEidt) [inline src/jit/compiler.h]: Customarily, members of `Compiler` that are associated with a particular phase or functionality are given a prefix (e.g. `fg` for flowgraph-related members). I'd suggest 'si' for these, though I believe that you're thinking about refactoring some of these out of `Compiler`, which would be good. These methods could easily be misinterpreted as being related to dataflow in the JIT.

File: src/jit/treelifeupdater.cpp [hunk 7]
```diff
@@ -168,6 +183,13 @@ void TreeLifeUpdater<ForCodeGen>::UpdateLifeVar(GenTree* tree)
                             VarSetOps::AddElemD(compiler, stackVarDeltaSet, fldVarIndex);
                         }
                     }
+
+                    if (ForCodeGen)
+                    {
+                        // BRIAN:: see how scopes are reported for promoted structs. Are they considered in the
+                        // amount of local variables to report?
```
Comment (CarolEidt) [inline src/jit/treelifeupdater.cpp]: This could be quite tricky, as I don't think there's a way to express, in the scope info, the notion that different fields of a struct live in different locations, e.g. if we had an instance s of a struct S that had 3 fields a, b, and c, we might have s.a in one register, s.b on the stack, and s.c in another register. I *think* that this would require modifying the format.

File: src/jit/codegencommon.cpp [hunk 8]
```diff
@@ -2349,6 +2368,25 @@ void CodeGen::genGenerateCode(void** codePtr, ULONG* nativeSizeOfCode)
 
     genSetScopeInfo();
 
+#if DEBUG
+    if (compiler->verbose)
```
[the remark this answers is not in the document]
Comment (BrianBohe) [reply src/jit/codegencommon.cpp]: It started just for debugging purposes, then I thought it would be useful to someone else. I will leave this without resolution so I can close it when the PR is about to be merged.

File: src/jit/compiler.h (hunk 5 above)
Comment (BrianBohe) [reply src/jit/compiler.h]: It's for printing only the VariableLiveRanges reported/used during a basic block on genCodeForBBList

File: src/jit/compiler.h (hunk 4 above)
Comment (BrianBohe) [reply src/jit/compiler.h]: Thanks!

File: src/jit/treelifeupdater.cpp (hunk 7 above)
Comment (BrianBohe) [reply src/jit/treelifeupdater.cpp]: How is it working now for debugging? It's all in the stack?

Comment (CarolEidt) [reply src/jit/treelifeupdater.cpp]: Right - we currently don't promote fields of structs in debug mode, just as we don't enregister any local variables.

**Comment by BrianBohe** [comment]
I have added some new flags to report variable live using either siVarScope or VariableLiveRange. VariableLiveRange is complete off when flag is not defined. But the other is just not being sent to the debugger. So I need to see the uses of siVarScope and put the flag around there too.

**Comment by BrianBohe** [comment]
I am splitting this PR to make it easier to review.

File: src/jit/compiler.h [hunk 9]
```diff
@@ -297,6 +297,199 @@ enum RefCountState
...
+
+    // Return an iterator to the first "VariableLiveRange" edited/added during the current
+    // "BasicBlock"
+    LiveRangeListIterator getStartForDump() const;
+
+    // Retutn wheter at least a "VariableLiveRange" was alive during the current "BasicBlock"'s
+    // code generation
+    bool hasLiveRangesToDump() const;
+};
+#endif
```
Comment (cshung) [inline src/jit/compiler.h]: // DEBUG

File: src/jit/compiler.h [hunk 10]
```diff
@@ -297,6 +297,199 @@ enum RefCountState
...
+
+    //  Make "LiveRangeDumper" instance points the last "VariableLifeRange" added so we can
+    // starts dumping from there after the actual "BasicBlock"s code is generated.
+    void setDumperStartAt(const LiveRangeListIterator liveRangeIt);
+
+    // Return an iterator to the first "VariableLiveRange" edited/added during the current
+    // "BasicBlock"
+    LiveRangeListIterator getStartForDump() const;
+
+    // Retutn wheter at least a "VariableLiveRange" was alive during the current "BasicBlock"'s
```
Comment (cshung) [inline src/jit/compiler.h]: Return whether

File: src/jit/treelifeupdater.cpp [hunk 11]
```diff
@@ -25,6 +25,14 @@ TreeLifeUpdater<ForCodeGen>::TreeLifeUpdater(Compiler* compiler)
 // Arguments:
 //    tree - the tree which affects liveness.
 //
+// Assumptions:
+//    The code corresponding to the given tree has been emitted, which means that if a variable has become live/dead
+//    or been spilled then the emiter is positioned after that changed. This is used to ensure [) "VariableLiveRange"
+//    intervals when calling "siStartOrCloseVariableLiveRange".
+//
+// Notes:
+//  If "ForCodeGen" is true, then the register mask use by CodeGen and the "LiveRangeList" of the "LclVadDsc"
```
Comment (cshung) [inline src/jit/treelifeupdater.cpp]: LclVadDsc -> LclVarDsc

File: src/jit/codegencommon.cpp [hunk 12]
```diff
@@ -684,6 +684,20 @@ regMaskTP Compiler::compNoGCHelperCallKillSet(CorInfoHelpFunc helper)
     }
 }
 
+//------------------------------------------------------------------------
+// compChangeLife: Update the GC's masks, register's masks and reports change on variable's homes given a set of
```
[the remark this answers is not in the document]
Comment (BrianBohe) [reply src/jit/codegencommon.cpp]: I have rephrased it [here](https://github.com/dotnet/coreclr/pull/23327), What do you think?

File: src/jit/codegencommon.cpp [hunk 13]
```diff
@@ -10640,6 +10679,47 @@ void CodeGen::genSetScopeInfo()
...
+
+                // a case that should only be considered for arguments
+                // The range may be 0 if the prolog is empty. For such a case,
+                // report the liveness of arguments to span at least the first
+                // instruction in the method. This will be incorrect (except on
+                // entry to the method) if the very first instruction of the method
+                // is part of a loop. However, this should happen
+                // very rarely, and the incorrectness is worth being able to look
+                // at the argument on entry to the method.
+                if (varDsc->lvIsParam && startOffs == endOffs)
```
[the remark this answers is not in the document]
Comment (BrianBohe) [reply src/jit/codegencommon.cpp]: Sorry for the delay, the range of variable home is intended to be close-open.

File: src/jit/treelifeupdater.cpp (hunk 7 above)
Comment (cshung) [reply src/jit/treelifeupdater.cpp]: Can we track this with an issue and a TODO comment?

Comment (BrianBohe) [reply src/jit/treelifeupdater.cpp]: Sure we can!
